### DISCUSSION TITLE
ROCKET_WORKERS proper setting?

### DISCUSSION THREAD

**Discussion by ghost** [dani-garcia/vaultwarden#3255] [Q&A]
I'm curious to learn more about the effect of setting ROCKET_WORKERS to higher values?
I currently have it set to 20, but i'm not really sure what that does?

Would increasing this value have performance gains for our users? Would increasing it too high have a negative effect? How would one find the right balance?

**Comment by tessus** [comment]
`ROCKET_WORKERS` sets the number of parallel threads that are used by rocket (the http request/response framework).

A CPU core has usually 2 threads, thus the OS will show you 2*cores processors available on that system.
This means that this number of threads can run in parallel by the CPU (socket) without having to context switch. (e.g. an AMD Threadripper with 16 cores will show you 32 processors, thus 32 threads can run in parallel.)

Now it depends on what other workload your system runs. There is just a rule of thumb (which is the default by rocket when not specified), but not a definitive answer. You will have to run performance tests to get those.

**Reply by BlackDex** [reply]
Besides that, depending on the OS, Kernel etc... there could be different scheduler types. Those determine which process gets cpu time, kinda a load balancer for the CPU. You can run 50 workers if you want, but they all need to wait for the CPU, Disk, Network and Memory (and multiple other resources) before they can execute the code.

**Comment by ghost** [comment]
Thanks for the feedback guys. Helps get me in the right direction I think.
We run vaultwarden in GCE on a dedicated VM (e2-highmem-2)
It has 2 vCPU's type Intel Broadwell
Currently has 20 ROCKET_WORKERS defined

We can see that overall CPU utilization of the cOS never really goes over 25% so there is room to spare there
![image](https://user-images.githubusercontent.com/48287057/219765455-31d00d79-daad-401a-bd43-0939e9831e1e.png)

We have had some end users complaining of slow load/search times (Also knowing that we are using the SQLite backend on local storage and the organization they are part of has ~ 3600 items) so I understand that there may be some other bottlenecks to address there, but i'm exploring anything we can do to assist in performance.

Given the fact that there are no competing workloads on this dedicate machine and CPU utilization is capping under 25%, It seems like I should be able to up the ROCKET_WORKERS threads to ~< 80 with current specs.... but even if increased to such a degree I would wonder if it would actually offer any sort of performance improvement. Or if each thread it just creating a longer "line" to wait for CPU time, would that actually cause a performance decline?

**Reply by BlackDex** [reply]
I would actually recommend to **not** set that env.
Let rocket/tokio handle it them selfs automatically.

If it doesn't perform, you could try to set it, but otherwise leave it empty.

**Reply by kpfleming** [reply]
> We have had some end users complaining of slow load/search times

Searching happens in the client, not the server. The server can't even look inside the contents of the data provided by the clients as it does not have the ability to decrypt it. If your users are complaining about long search times in their clients, that will have to be addressed in the clients or the machines they are running on.

What does 'load' mean in this context?

**Reply by tessus** [reply]
Nope, this is not necessarily the case. You _could_ have already reached some sort of contention at 25%. This means that no matter how many workers you set, the CPU will never go above 25%. But you have to look at the I/O wait times, system CPU usage in addition to the user CPU usage, and so on.

**Reply by BlackDex** [reply]
Also, how many ciphers do those users have access to. That could impact a lot if that is high.

**Comment by ghost** [comment]
This is very enlightening. I did not even think about the client side portion of this. Extremely good point.

In terms of "loading" I have seen two things happen as of late. When users access the web vault, sometimes it takes between 3-10 seconds for the page to render a list of the items. The outline (top menu, organization/folder list) rendered quickly but the item list is delayed very noticeably, and more so when searching for items there is a significant delay. I attribute this is part to the large number of items in the effected users scope of access of course as well.

The other scenario I have seen is that a few users have logged into the webvault and just received this page.
This user does have a number of items in their vault as well as a member of multiple organization which do not appear in the sidebar. In these cases however clearing the cache of their browser seems to fix it consistently so we haven't worried about it too much given the low refrquency of reports and the simply enough fix. Not performance related exactly, but perhaps worth mentioning.

![image](https://user-images.githubusercontent.com/48287057/219778158-51ce8b5d-f135-4a2f-8966-1e641a2a08e7.png)


The GCE platform is unfortunately fairly metric poor in terms of monitoring IOPS and I/O wait times. One of my complains around the platform. I will see if i can dive deeper into the platform to gather more information in that regard.

**Reply by BlackDex** [reply]
Sounds to me like a huge vault.
In the browser, use F12 and check how long the sync call takes.

**Reply by tessus** [reply]
> The GCE platform is unfortunately fairly metric poor in terms of monitoring IOPS and I/O wait times. One of my complains around the platform. I will see if i can dive deeper into the platform to gather more information in that regard.

But you can ssh into the VM and use OS tools to drill down into the perf issues. Start [here](https://netflixtechblog.com/linux-performance-analysis-in-60-000-milliseconds-accc10403c55) and go from there.

**Reply by ghost** [reply]
@BlackDex 

Appears my longest time is on the websocket. Forgive me as I am not a web developer and i'm not the most skilled in using and interpreting the developer tools here.
![image](https://user-images.githubusercontent.com/48287057/219787878-e8f86b0e-25e5-4352-9fa4-33455219df1c.png)
![image](https://user-images.githubusercontent.com/48287057/219788040-73a323fd-9030-4e16-8f04-33cb2373e2b8.png)

**Comment by jjlin** [comment]
3600 items is a pretty large amount, and in my experience, the bottleneck is the client parsing (and indexing?) the huge JSON document that represents the vault during a sync. If you aren't running the `testing` branch, full vault syncs also occur much more often than they should -- when you activate the browser extension, it often checks whether it should do a full sync, which then blocks the UI until the download and parsing finishes. This is what #3124 fixes. I suspect that's likely the cause of the reported slow load times, as I've seen that myself prior to the fix.

**Comment by ghost** [comment]
@tessus Thanks. I appreciate the link there. Looks to be a comprehensive resource. I will try to get ssh access to the host (currently protected by some bureaucratic security SOPs in my organization i will need to hop through). I know the Google "Container Optimized OS" that this runs on is in no way full blown linux and i'm not sure it will have the packages present (or installable even) to perform some of these functions but i will give it my best college try.

**Reply by tessus** [reply]
You are welcome, although it's just a quick perf overview of the system. If you find something, you'll have to drill down and get to the bottom of it. Ask a perf engineer in your company to help with profiling and more complex tasks, when you narrowed it down a little.

But jjlin has a point, I think you should try the latest testing branch that has the perf sync fix.

**Comment by shoaib42** [comment]
@dwestness 

My 2 cents here,

> 2 vCPU's type Intel Broadwell

Sure you get this, but a Xeon Brodwell starts with 4cores + hyperthreading = 8 vCPU (best case scenario) ===> think noisy neighbour. I wouldn't go with this VM for a CPU intensive workload.

Your CPU profile of 25% doesn't tell much, if you can as dear @tessus has pointed, check vmstat on the instance. I'll be looking for the `b` and the `st` column. Don't get carried away by `cs` (context switch) as that only helps when we have something to compare. You need to have `b` at 0, if you see `st` > 0 then there is some crazy party going on at the socket. Check for `wa` the io-wait, best to have this at zero.

Also check the `txqueuelen` , if you have a lot of traffic you need to increase this over 10k, usually you get a juicy 1k which never is good for a server and very vanilla. Check the `somaxconn` values, you may need to increase (try 512 or higher 2048 maybe), this could very well be your bottleneck (your requests may be just sitting queued to be accepted)

**Comment by BlackDex** [comment]
The server side is not there issue i think.
You could test the `testing` tagged images which have some small server side optimizations. But also a newer web-vault which also should have some performance improvements with large vaults.
